lib/std/Thread/Pool.zig

```diff
@@ -75,6 +75,65 @@ fn join(pool: *Pool, spawned: usize) void {
     pool.allocator.free(pool.threads);
 }
 
+/// Runs `func` in the thread pool, calling `WaitGroup.start` beforehand, and
+/// `WaitGroup.finish` after it returns.
+///
+/// In the case that queuing the function call fails to allocate memory, or the
+/// target is single-threaded, the function is called directly.
+pub fn spawnWg(pool: *Pool, wait_group: *WaitGroup, comptime func: anytype, args: anytype) void {
```

Comment:
Thoughts on generalizing this into a scoped task API instead?
```zig
pub fn scope(pool: *Pool) Scope {
    return .{ .pool = pool };
}

pub const Scope = struct {
    pool: *Pool,
    wg: WaitGroup = .{},

    pub fn spawn(self: *Scope, comptime func: anytype, args: anytype) void {
        self.wg.start();

        const Args = @TypeOf(args);
        const wg_func = struct {
            fn complete(wg: *WaitGroup, func_args: Args) void {
                @call(.auto, func, func_args);
                wg.finish();
            }
        }.complete;

        self.pool.spawn(wg_func, .{&self.wg, args}) catch wg_func(&self.wg, args);
    }

    pub fn join(self: *Scope) void {
        self.pool.waitAndWork(&self.wg);
    }
};

// var scope = pool.scope();
// defer scope.join();
//
// scope.spawn(doWork, .{with_args});
```

Reply:
for simple scoped CPU-heavy tasks, I really love Sorbet's API

https://github.com/sorbet/sorbet/blob/bf58b629381a0e1268a3b9dc8034cca8bd684ffc/common/concurrency/WorkerPool.h

The trick of that API is that, instead of submitting _one_ job to a single thread in the pool, the job is submitted to _all_ threads at the same time. So, rather than doing something like

```zig
for (tasks) {
    pool.spawn(task, struct {
        fn runTask(task: *Task) void { ... }
    }.runTask)
}
```

you do

```zig
const queue = ConcurrentQueue(Task).init(&tasks); // a pair of atomic counter and a slice
pool.multiplex(queue, struct {
    fn run(queue: ConcurrentQueue(Task)) void {
        while (queue.pop()) |task| runTask(task)
    }
}.run);
assert(queue.slice.len == 0);
```

I should probably blog about this --- a nice pattern which I haven't seen anywhere else

Reply:
It looks to be creating a mini thread-pool inside the thread pool task. A down-side is that it requires having all tasks contiguous or exist/be-allocated at the same time. A plus-side is that your tasks are scheduled LIFO regardless of the pool's internals.